### PR TITLE
Add refresh token expiry hook

### PR DESCRIPTION
**Pull request by zacps**
This adds a hook which is used to determine whether an access token is still valid, or should be expired.

The default implementation never expires refresh tokens which (mostly) matches behaviour of public OPs.

In the event a refresh token is expired it is immediately removed from the database, along with the access token.

* Subsequent requests to the token endpoint with `grant_type=refresh_token` will return `invalid_grant`.
 See [RFC6749#5.2](https://tools.ietf.org/html/rfc6749#section-5.2)
* Subsequent requests to the token introspection endpoint will return `{'active': false}`. 
See [RFC7662#2.3](https://tools.ietf.org/html/rfc7662#section-2.3)
* Subsequent requests to routes behind `protected_resource_view` will return HTTP 401 unauthorised and send the `WWW-Authenticate` header. 
See [6750#3](https://tools.ietf.org/html/rfc6750#section-3)

An `issued_at` field is added to the token model and passed into the hook to make it possible to implement a simple time-based expiry. This field is not used elsewhere in the application.

---

Alternatives:

This approach, while flexible, leaves a lot to the consumer of the provider. It may be more useful to provide fixed options for how refresh tokens should expire.

Deleting tokens when they are no longer valid is different to django's session management. It might be more idiomatic to keep 'dead' tokens in the database and add a CLI command `cleartokens` to remove them.

### CODE REVIEW

File: docs/sections/settings.rst [hunk 1]
```diff
@@ -267,4 +267,13 @@ It is recommended to set this to ``'filesystem'`` for new applications if possib
...
+    need to sync RSA keys between your application servers.
+
+OIDC_REFRESH_TOKEN_ALIVE_HOOK
+=============================
+
+.. currentmodule:: oidc_provider.settings
+.. autoattribute:: DefaultSettings.OIDC_REFRESH_TOKEN_ALIVE_HOOK
+
+.. currentmodule:: oidc_provider.lib.utils.common
+.. autofunction:: default_refresh_token_alive_hook
```
Comment (zacps) [inline docs/sections/settings.rst]: I'm not sure I'm happy with how this documentation renders. It does have the advantage of not having to duplicate the docs between `settings.py` and `settings.rst`.

**Comment by zacps** [comment]
The `cleartokens` command is the approach taken by [Django OAuth Toolkit](https://django-oauth-toolkit.readthedocs.io/en/latest/management_commands.html#cleartokens).

File: oidc_provider/tests/cases/test_token_endpoint.py [hunk 2]
```diff
@@ -446,6 +449,36 @@ def do_refresh_token_check(self, scope=None):
...
+        post_data = self._refresh_token_post_data(token.refresh_token)
+        response = self._post_request(post_data)
+        self.assertIn('invalid_grant', response.content.decode('utf-8'))
+
+    @override_settings(OIDC_REFRESH_TOKEN_ALIVE_HOOK=Mock(return_value=True))
+    def test_alive_refresh_token(self):
+        """
+        Test that a dead refresh token, as defined by the hook, is rejected with 'invalid_grant'.
+        See https://tools.ietf.org/html/rfc6749#section-5.2
+        """
```
Comment (adamlynam-catalyst) [inline oidc_provider/tests/cases/test_token_endpoint.py]: If I am following correctly, I think this test has the wrong comment.

**Comment by adamlynam-catalyst** [comment]
Not part of your changes, but I noticed that `TokenEndpoint.validate_params()` method is very long. I am sure it all works correctly, but that is a lot of code to grok at once.

Definitely out of scope for this merge, but might be something worth refactoring if we touch it again.

**Review by adamlynam-catalyst — CHANGES_REQUESTED** [review]
> In the event a refresh token is expired it is immediately removed from the database, along with the access token.

> Deleting tokens when they are no longer valid is different to django's session management. It might be more idiomatic to keep 'dead' tokens in the database and add a CLI command `cleartokens` to remove them.

I agree that being consistent in approach is worthwhile. However, for now, deleting the tokens is fine; we can revisit it another time to bring it in line with the Django paradigm of deleting through an offline operation later.

> This approach, while flexible, leaves a lot to the consumer of the provider. It may be more useful to provide fixed options for how refresh tokens should expire.

The implication from the spec seems to be that refresh tokens work indefinitely, unless revoked. I think this hook provides flexibility to expire them another way and anticipating that would probably leave us falling short.

>     * Subsequent requests to routes behind `protected_resource_view` will return HTTP 401 unauthorised and send the `WWW-Authenticate` header.
>       See [6750#3](https://tools.ietf.org/html/rfc6750#section-3)

Is it okay that if the refresh token is expired that the connected access token stops working too? I understand the two are conflated in the provider implementation and it seems like in all practical implementations this is desirable. I can only think of a very hypothetical situation where refresh tokens are immediately expired and the original access token issued together with it immediately stops working too. Perhaps that would be unexpected or unwanted behaviour?

^ This last one might do with an in person chat, I may just be totally misunderstanding the expectation here.

Above just discussion, only actual change required is the test description.

**Comment by zacps** [comment]
Authorisation providers are [allowed to chose](https://tools.ietf.org/html/rfc6749#section-1.5) when to issue a refresh token. I think that'd handle the case when you don't want a valid refresh token to exist but want an access token.

Currently the provider unconditionally issues refresh tokens, adding a hook to determine whether to issue one is something I've considered.

**Comment by adamlynam-catalyst** [comment]
> Authorisation providers are [allowed to chose](https://tools.ietf.org/html/rfc6749#section-1.5) when to issue a refresh token. I think that'd handle the case when you don't want a valid refresh token to exist but want a access token.
> 
> Currently the provider unconditionally issues refresh tokens, adding a hook to determine whether to issue one is something I've considered.

I agree, that sounds a much better way of handling the situation. Perhaps something we can add to the backlog, doesn't feel like it should be in the scope of this change.

**Review by adamlynam-catalyst — APPROVED** [review]
LGTM

**Pull request merged**